File: src/lib/sources.ts
import { existsSync } from "node:fs";
import { dirname, resolve } from "node:path";

export function resolveOverridesPathFromSourcesDir(sourcesDir: string): string {
  const localOverridesPath = resolve(sourcesDir, "overrides.json");
  if (existsSync(localOverridesPath)) {
    return localOverridesPath;
  }

  return resolve(sourcesDir, "../overrides.json");
}

export function resolveHintsPathFromOverridesPath(overridesPath: string): string {
  const overridesDir = dirname(overridesPath);
  const localHintsPath = resolve(overridesDir, "hints.json");
  if (existsSync(localHintsPath)) {
    return localHintsPath;
  }

  return resolve(overridesDir, "sources/hints.json");
}

File: internal/pipeline/sources.go
package pipeline

// sources.go ports src/lib/sources.ts path resolution helpers verbatim.

import (
	"os"
	"path/filepath"
)

// ResolveOverridesPathFromSourcesDir mirrors resolveOverridesPathFromSourcesDir:
// <sourcesDir>/overrides.json when it exists, else <sourcesDir>/../overrides.json.
func ResolveOverridesPathFromSourcesDir(sourcesDir string) string {
	local := filepath.Join(sourcesDir, "overrides.json")
	if fileExists(local) {
		return local
	}
	return filepath.Join(sourcesDir, "..", "overrides.json")
}

// ResolveHintsPathFromOverridesPath mirrors resolveHintsPathFromOverridesPath:
// <dir>/hints.json when it exists, else <dir>/sources/hints.json.
func ResolveHintsPathFromOverridesPath(overridesPath string) string {
	dir := filepath.Dir(overridesPath)
	local := filepath.Join(dir, "hints.json")
	if fileExists(local) {
		return local
	}
	return filepath.Join(dir, "sources", "hints.json")
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}
